lib/msf/util/dot_net_deserialization/formatters/soap_formatter.rb

```diff
@@ -0,0 +1,96 @@
+module Msf
+module Util
+module DotNetDeserialization
+module Formatters
+module SoapFormatter
+
+  class SoapBuilder
+    def initialize(stream)
+      @document = Nokogiri::XML::Document.new
+      @root = envelop = node('SOAP-ENV:Envelope', attributes: {
+        'xmlns:xsi'              => 'http://www.w3.org/2001/XMLSchema-instance',
+        'xmlns:xsd'              => 'http://www.w3.org/2001/XMLSchema',
+        'xmlns:SOAP-ENC'         => 'http://schemas.xmlsoap.org/soap/encoding/',
+        'xmlns:SOAP-ENV'         => 'http://schemas.xmlsoap.org/soap/envelope/',
+        'xmlns:clr'              => 'http://schemas.microsoft.com/soap/encoding/clr/1.0',
+        'SOAP-ENV:encodingStyle' => 'http://schemas.xmlsoap.org/soap/encoding/'
+      })
+
+      body = node('SOAP-ENV:Body', parent: envelop)
+
+      stream.records.each do |record|
+        record_value = record.record_value
+        case record.record_type
+        when Enums::RecordTypeEnum[:SystemClassWithMembersAndTypes]
+          build_class_with_members_and_types(body, record_value)
+        when Enums::RecordTypeEnum[:ClassWithMembersAndTypes]
+          library = stream.get_object(record_value.library_id)
+          build_class_with_members_and_types(body, record_value, library_name: library.library_name)
+        end
+      end
+    end
+
+    attr_reader :root
+
+    protected
+
+    def build_class_with_members_and_types(body, record_value, library_name: 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089')
```

Comment:
Would it be worth having some constants that contain well-known, pre-defined libraries with versions and pubkeys?

Reply:
This is a really interesting question, it's kinda related to your other comment:

> Should this library name be configurable for other .NET versions? OR is 4.0 the only one supported at this time? I'm guessing the latter because it's the earliest version this issue appears in?

I agree, I hadn't quite realized how often this string appeared in all the chains, especially the TypeConfusion one. To address that I started a new module in commit 6995a9a775af4d311aae9e49aade8b6dee660d3a that aims to add the plumbing to support different versions in the future using the `Assemblies::VERSIONS` hash. I left the Powershell Editor one out of it because I'm honestly having a hard time identifying with 100% certainty if the 3.0.0.0 is the library version of the .NET version :sweat_smile: . As for trying older .NET versions, it should be relatively easy to test them now by generating the Strong Name details with [sn.exe](https://docs.microsoft.com/en-us/dotnet/framework/tools/sn-exe-strong-name-tool), I just need to do this and make sure the gadget chains will still work.